Local notification: Notification to purchase plan 6 hour after trial subscription

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #10264 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds a new scenario for local notification to remind users to purchase a paid plan 6 hours after their trial subscription.

Since we are replacing notification scenario 5 with scenario 7, this notification uses the same wording as scenario 6.

Since the logic for scheduling post-trial subscriptions is very similar for scenarios 6 and 7, I added some refactoring to reuse the logic with a method named `schedulePostSubscriptionNotification`. We just need to specify the scenario and the time interval for scheduling the notifications and the rest are taken care of.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Build and run the app on a physical device for easy testing.
- Create a new store with a free trial subscription.
- When the store creation completes, notice in Xcode console: `🔵 Tracked local_notification_scheduled, properties: [AnyHashable("blog_id"): <site_id>, AnyHashable("is_iap_available"): true | false, AnyHashable("type"): "six_hours_after_free_trial_subscribed<site_id>", AnyHashable("is_wpcom_store"): true]`.
- Open the device Settings app to switch to 5h58m after the time of the store creation.
- After 2 minutes, a notification asking for purchasing a plan should be displayed.
- Tap the notification, the app should be opened. Notice in Xcode console: `🔵 Tracked local_notification_tapped, properties: [AnyHashable("type"): "six_hours_after_free_trial_subscribed<site_id>", AnyHashable("is_wpcom_store"): true, AnyHashable("is_iap_available"): true | false, AnyHashable("blog_id"):<site_id>]`.
- The app should navigate to IAP if you are testing with US storefront, or the plan detail screen otherwise.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/3f6df7ed-8e85-4d3b-9055-0a3d354c05ad" width=320 />

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Add details for new notification scheduled 6 hours after trial subscription

itsmeichigo committed:
Update test for local notifications

itsmeichigo committed:
Handle tap on scenario 6 notification

itsmeichigo committed:
Add test for handling tap on scenario 6 notification

itsmeichigo committed:
Schedule 6 hour post subscription notification

itsmeichigo committed:
Add tests for scenario 6 notification

itsmeichigo committed:
Merge branch 'trunk' into feat/10264-notification-6hr-after-trial-subscription

itsmeichigo committed:
Cancel notification and fix build failure

itsmeichigo committed:
Update test for AppCoordinator

itsmeichigo committed:
Update release notes

itsmeichigo committed:
Remove redundant variable

itsmeichigo committed:
Fix typo in release note

itsmeichigo committed:
Update test for test_local_notifications_are_canceled_after_the_site_is_upgraded

selanthiraiyan reviewed (commented):
The event's `type` property has the siteID appended to it. `six_hours_after_free_trial_subscribed123456`

```
Tracked local_notification_scheduled, properties: [AnyHashable("is_iap_available"): true, AnyHashable("is_wpcom_store"): true, AnyHashable("blog_id"): 123456, AnyHashable("type"): "six_hours_after_free_trial_subscribed123456"]
```

I think we also have to update the `LocalNotification.Scenario.identifierForAnalytics` method to avoid this.

itsmeichigo committed:
Update identifier for analytics

itsmeichigo committed:
Fix test failure for test_local_notifications_are_canceled_after_the_site_is_upgraded

itsmeichigo committed:
Update comment for test_local_notifications_are_canceled_after_the_site_is_upgraded

itsmeichigo committed:
Increase timeout for notification tests

itsmeichigo committed:
Add missing inAppPurchaseManager to StorePlanSynchronizer in the failing test

itsmeichigo commented:
> I think we also have to update the LocalNotification.Scenario.identifierForAnalytics method to avoid this.

Good catch, I updated this in 75d08b4caaa746fcf673deda11ff2b51b778ee80. I also fixed the failed unit test in 52affc3f7927e0a63713e76242773839adc11fe9. This PR is ready for another look, thanks!

itsmeichigo committed:
Merge branch 'trunk' into feat/10264-notification-6hr-after-trial-subscription

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Upgrades/StorePlanSynchronizer.swift:
```diff
@@ -135,30 +147,36 @@ private extension StorePlanSynchronizer {
-    func schedule24HrsAfterSubscribedNotification(siteID: Int64, subscribedDate: Date) {
-        /// Scheduled 24 hrs after subscribed date
-        let triggerDateComponents = subscribedDate.addingTimeInterval(Constants.oneDayTimeInterval).dateAndTimeComponents()
+    func schedulePostSubscriptionNotification(scenario: LocalNotification.Scenario,
```
Thanks for adding this reusable method!

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Upgrades/StorePlanSynchronizerTests.swift:
```diff
@@ -193,16 +257,26 @@ final class StorePlanSynchronizerTests: XCTestCase {
         // Then
         waitUntil(timeout: 3) {
-            pushNotesManager.requestedLocalNotificationsIfNeeded.count == 1
+            /// 2 notifications include:
+            /// - 6 hrs after trial subscription
+            /// - 24 hrs after trial subscription
+            /// Update this number based on the number of notifications we support.
```
Nice work adding these comments. 👏

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Upgrades/StorePlanSynchronizerTests.swift:
```diff
@@ -193,16 +257,26 @@ final class StorePlanSynchronizerTests: XCTestCase {
         // Then
-        waitUntil(timeout: 3) {
-           pushNotesManager.canceledLocalNotificationScenarios.count == 3
+        waitUntil(timeout: 5) {
```
❓ Do we have to increase the timeout here?

itsmeichigo replied:
Since we have 4 different kinds of notifications with async-await, I saw that the test seems flaky occasionally. This timeout only ensures that the test waits longer if the conditions are not met. Otherwise, it can still finish sooner.

selanthiraiyan reviewed (approved):
🚀
